### ISSUE TITLE
Current status for granular (single component) redraw?

### ISSUE THREAD

**Issue by Niklas81** [MithrilJS/mithril.js#2388]
Having spent a long time building a project skeleton to suit my needs, I just realized there is no easy or build-in way (or hard way) to perform a redraw that affects only a single component, which really comes as a punch in the face.

I've looked through several old issues, all with long discussions and ideas for various ways of solving this, but I cannot find anything solid that makes this possible. Userland solutions are fine with me, altough i would prefer something as simple as `m.redraw(this)` or `m.redraw(false)` or `m.redraw('MyComponent')`.

[Mithirl Island](https://github.com/barneycarroll/mithril-islands) might have have been useful, but what I can gather, it won't work in Mithril >= 2.0.

Related issues are: #1907, #1847, #1765, #1776, #1165, #1837, #1154

#1154 pinpoints my need for this quite well:

> The component that cause the issue is a notification container. This component receive notifications from websocket events (specifically firebase). It means at anytime this component can receive a new notification.

This sort of feature must be quite common, so I really feel this should be supported in a sane, simple way.

So, how can I do this in a future-proof way; that won't stop working in future releases of Mithril?

**Comment by cavemansspa** [comment]
@Niklas81 -- there was a [related discussion](https://gitter.im/mithriljs/mithril.js?at=5c65cf0e4bafd97ca3f1c82e) in gitter a few weeks back.

this [gitter message](https://gitter.im/mithriljs/mithril.js?at=5c65dbe2604f233ab60ecc87) provides an [example flems](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4hcYgAIAysQzEArnABCGAE6SAvJIAUASi0A+ScAA6aSZNhSA+gBNaWLZIVo7MSGhh3zl8VIl5GGczC0tJOldiREkARgpfSVZzRLUYRTULUPDJemo0oJidYAcsVgNNY2ycyXtHZ1LEnLh0gElhNQA3DCgdfSMTJprAxnxIhmcRmDHaKMkAajiAbiGcgHo1rHw0uzUMAHd9FbCayS20txg1HTqsCkkAAwASYCmZqNZJYggcOEl5SQvHRefaSAAiQX0enwxFoABlaNQejAACo-GCyNQQNAAc30rAeelWSXusQADBS9MdLOxJENOhAYPsihVjFgdAByOwQTocolhVjHQUpND+STtAAOCmIyloAA9nP1KoMwmKpiF6T0FDAYhyABIwKBQWiSfa0NRQOwchIC1LpBSZFU5BlMlkDdkc7FS4jIYgATwlME0jDlxAAutaTHSTuFulBtTE3nHtTbTvQvdKip1WZIk1rgtpOjD1Dj0vhkzBU+FyolkmhhWhzGKAAoYUuKnPFSQu5m6HPIdmS6WyuV6e7s2TyJSqNR6MPlY7mLZYWYMHQOagKHAMfAAI1odj991bpaplBALVg1G+9AQPFiACZEAA2MlsDggTA4PBjOACGj0IwzA8GwYZUFA2IANZ3qgn5cHgWAQMQhBYuQVAOuQPAkMQEpwIgGyuBKkE4jMWCbEhKHQAAArE+C0c+5HIah+D8Oe-qBngcD5BAEqiKw85AA) that demonstrates one to approach to a self drawing component.

**Comment by osban** [comment]
As you need a mount point, `dom`, or `vnode.dom` represents the dom of the component after it has been created. You could also do something like [this](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4A9GIAE9KAE9JAJxgATBRgDuk4oRiS4xDMQCucSQCMMCgDpoh+yQGUDxuACFLkgLySAFAEovAD5JYBtJSVhiSTojBi9JAAYwyWSlYwU0EOTw+molQxhEXwDPYNDM8Mq4GGIASWEFADcMKB9-IKyKysqYhgBqPuzuySx8JTRlGAUfRn0KSQADABJgXuJWLQgcU0NJFZ80GE0AEQL-P3xiWgAZWmoWmAAVLZgnBQg0AHN-VgW-IfC7EkAEYEmD-l1JECho0IEciu1SiMfAByZQQRoAYlmxBREMBNlYNhsdiiGAADuT4oiyslIloYAAPKLeFEACRgUCgtEk6loCigyhRyVSNSMGU63Vh8OKQQByLRGJRFHl4SwqI+5KMuPmwBSkOGkmaUCMhQZzJVBuG9E12qKuiRjGZ8Rgl0snxq+GNptVUL8lsNap8TkMJncCnxw0jRLQMZsoywtFixB8yjuRhwDHwZloylk8wp5IhlBA1Vg1GIEHoCB4AGZgYgEmwOCBMDg8PhqHABDR6IxmDw2ABdKhQD4AaxrqFbXDwWAg2ne5Co4vIPBIxHJcEQEli5PHn07tCwYnni+gAAFgfhrwA2U8LwhL-D8EvEWTk7ilvIQcmiVhDqwQA).

**Comment by Niklas81** [comment]
So if I use `m.render(test` I'll mount it on a any element with the id `test`..? I was wondering about the weird object construct `{dom}`; an object with an attribute name but no value, what are they called? Can't search for it without knowing what it's called! :relaxed:

**Comment by Niklas81** [comment]
The most pressing thing that needs to work, is attaching an eventlistener inside this magic component, how can this be done?

**Comment by osban** [comment]
Yes, it's similar to `document.getElementById('test')`, so you could also use that. `dom` is deconstructed from `vnode` (look [here](https://mithril.js.org/vnodes.html) for more info), just like you could do `view: ({attrs}) => {...}`. Where/how do you want to use the event listener?

**Comment by Niklas81** [comment]
For at least two things I can think of right now:
* One needs to listen to events sent through websockets. Either directly, or I will set up this outide of mithril components (so I can maintain same connection without having to reconnect for every page rendering; I'll have to look into this more) and from that emit a custom event.
* The other will listen to progress events on download of data. Most likely it will be a custom event that aggregates multiple concurrent downloads into one percentage-digit.

**Comment by Niklas81** [comment]
@isiahmeadows : if I use `m.mount` instead of `m.render`, will I then be able to create a full component rather than just a view?

**Comment by cavemansspa** [comment]
@Niklas81 -- there's a progress bar example by @spacejack that's `m.render()` based. [flems link](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4A9ACoAOmhkAVQhDgACAOQB5AEZwYAJwBue1crpYADvRHKYADzMxqxFcUIxlGYsV0qAJvGq6EJowPsrBULQA7vjyrsr6PrRYyrpMfqmhmgCeygAG+hAwkbnKkdBQKTAAtKlo6cpotLpYGFBQOWUuELJoALIQLoFQqioZuhiRJlnUsDFoCm4JSZV1eiFhObm16SVlbcr07ZU1aXqlrmjKLjAy+q0Arm5wXjDYyvdmPh7w+MoKSqVyg1aMQroEAObgs7giKaVqVHzjSJwOZbU66ErbPQqbrEWhXOKJZKNPzKPxwAJBdbZPI6WCOJq5CgyEGuXRlHRkmBgDD3KBOK7464mJIWNAiEYpWgg4F+OYyLQ6Ax6ACCnm8AB45AA+ZTAGRSYh3KCPRDKADKL2wWu1zNkThg9LxugA-GbnoE0OC7YaCkUABT6EkwM0ANWDAEow8GfS9VrpAw8Q38o8pw7Q-DJWAa0BIxDIhM8Dto9IZdMoALzKMD3NCOCD0f3ADxeOBm4DG03xJPm3TUVisCN6mTKZRiMTKADCqW+7i59lWDGUHteyUibmoGEu9zg920lOCgoOl1SWFoxpHIrQRc7birt971HwLTMgcruv0EcvqWI910l31S5R2PAJXkYZQmyJCh3HVNs9SxXRoLpBxnQHIdAOA4Dx2UAARCBQnXZQoVBDBl0dFCmgOcs4EIWg+QItwEKPXRpVBYMXQNYhDUNQtQUdSsriyexaDAMinUoitJLUD1unBVROO4rjiBdMkknwABHR5dCyc1yIZBNkP0r97UUw0zSJBSlOwgAlaomMOHJ6GXK1kg+L5GDgSzfSTZ8MFffR32HEylMNLB8AQ-1HWgiLP2M0ziEHS9R3YJLj1Pc8YH9dDUtHbCAFVrz3ClAkPMAWOSFdsBy7sTRgfA0n9LxHiHZRUpSoDRz9SJA0jILMNHH8-0uIMMzqltvHwLqetG4zMOzID5vmmRJGkS4JD+OIIgwHxZIAIQwctTDFKxCC3HwVBE0TKoqoqDxgMRa13fcSpuNblBaTAoRwJczverd7nhJi8QJNxaBLZVDtFSwGDmZR1oAdTcPwZgOxZCkmf06MYXQh1+pj-W6cUcerWt63oZxTq4t7hTBpUyxFcxodBPYKm6fRaAAazqkc8wLcnQS2navQEwDfXRiDm1g9szBY8FUjgOA0KCxTBv-d7MfBstoIw-rsIWMSKPLHd1mByA6hBlZ6mIA7iOsWBvqp-qDf0s1VHwOFdCqLGjFjRTbzNGXaDl+A4F9pT8cD4OFey4KVZgX81awf03Y91RteeLJYDNXIyh8FwzQAEmASP5ZUdaAEYAAZK9YABSXJEtjhKw7FgMY-iw1VcuJO3cF2Sqk0CJqA5tPqswnuzDT5QS5DrLlA1ZRy+UFSpBAAAZWhttk-Ad9X5QzVXnDLAAQlXiM7SdzC9bia3dFt8iHeY9ELZnhWwgO5RcVoMfgJ792Dq9vcbGqhZqXziiFBKllG6GkWjmSQyhegYG6O4MwZgGbHSXLzOs-MUFoKrDrKaMdMJd3Vm7Pyk8L79R7oQcuU9V69ByCqVBZ9KHj39H3L02tX6K3PtVbhc9tRViXgAMmEaQihP8e6aCAXiNAo8OqXx1pfYCxAhLJlUNIzw9B5HKMwvQGYEBh7ulvsQVhl92q6NaiAWyzwDpcRAD-UBmEnHKFAbAno2EAAStAoChG4c5WcyDSLXRkP4qsYVrpZTgROc0EAsB8lnBw8EgIXBzmsiqAAYsoCItAzAyBrHWYgDZLiCw3rkueOt+H8KHAAamUJXfA1cACsoCICiX9NU+ei8iHAVSJpeAxAVRoDiR4YpGTxg4HYZvHwZSzBuLtpyJRY4JxHxJvsMYExzhMGydMkIqUwobO6m4rMOYClk0uLY3QxAKmXn4ZXUBfTHjPCGSMop9BxlcCmdtWZxklrXhMVEjxE5eh0SXMKchMgwpnlrNcxI1B7gO3dhmLI0FyHGTYBwEAmAcB4Hw)

**Comment by isiahmeadows** [comment]
@Niklas81 You can create a full component by just rendering `m.render(vnode.dom, m(Comp, ...))`. In fact, `m.mount` internally does `m.render(root, m(Comp))`.

**Comment by osban** [comment]
@Niklas81 you can use event listeners just like normal, but I think the easiest way is to couple your data to your view, use `m.mount`, and `m.redraw` when the data changes. There is an example of attaching event handlers to the xhr in `m.request` [here](https://mithril.js.org/request.html#monitoring-progress).

@cavemansspa your flems link is borked

**Comment by Niklas81** [comment]
@cavemansspa thanks for the link, I'll look into it! My usecase is different as I use Axios which has an ondownloadPrgess method attached to it, which I will hook into. And I will have multiple concurrent downloads that I will need to aggregate.

@isiahmeadows ok cool!

@osban the whole purpose of doing this in an isolated component is to **not** redraw the whole page whenever there is a notification; thereby not resetting e.g. textareas that users have started filling. 

> you can use event listeners just like normal

How **do** you do this? Normal is `document.getElementById('someElementId').addEventListener.('myCustomEvent', function(event) { console.log(event) }` but how and where do I add this with Mithril?

**Comment by osban** [comment]
@Niklas81 I added one to my earlier [example](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4A9GIAE9KAE9JAJxgATBRgDuk4oRiS4xDMQCucSQCMMCgDpoh+yQGUDxuACFLkgLySAFAEovAD5JYBtJSVhiSTojBi9JAAYwyWSlYwU0EOTw+molQxhEXwDPYNDM8MrGfXwMZWUAUQA3EQAZCH0mGAUfAHJqKAhqAGteimKgyQAHHwADWiwAc0kASV6W8xgmaMGRlQBCWb8-bMq4GGIV4QUmjCgff0nyypfo2ljiAGpP05esfCUaGU3R81WI41mABJgDEGKwtBAcKZDJJoT40DBNAARAr+Pz4Yi0Vq0ah3GAAFURMCcCggaEW-lYR1+knYkgAjAkuScKuE2acmhBMUVHqVJFg+soIE0AMRg3o8vk2Vg2Gx2KIYKZTeKisrJSJaGAADyi3l6AAkYFAoLRJOpaAooMpeslUhcjBksrzJILhRNSiyJb0pU0xizwkG6VMjMQxiEUt7Xj67kZCoaTRRw5V6FGY0VdGLGCb4jACZZFhd8LcoKms6w-JnE5UJU5DCZ3ApFa8uyq0L2bP8sO8GD5lCSjDgGPgzLRlLJxpqpjzKCBzrBqMQIPQEDwAMwAJkQCTYHBAmBweHw1DgAho9EYzB4bAAulRBmhhjvUGeuHgsBBtFpcgqA9cgeBIYgpjgRAJFiKZhkWK8FjEf9AOgAABdl8CwgA2FCAMIID8H4FdiFkKZuFXPIICmURWGfVggA).

**Comment by Niklas81** [comment]
Awesome, thanks! Should I consider reasons not to use an element with an id to emit it onto, or is that the best way to go? I mean, `document.getElementsById('myEventReceiver')dispatchEvent(event)` vs `document.getElementsByTagName('aCustomTag')[0].dispatchEvent(event)` vs **[something else, if so, what?]** ??

**Comment by osban** [comment]
I don't know, but my guess would be that it doesn't matter.
As you can see [here](https://flems.io/#0=N4IgZglgNgpgziAXAbVAOwIYFsZJAOgAsAXLKEAGhAGMB7NYmBvEAXwvW10QICsEqdBk2J4hcYgAIMABxmSAvJIAUASkUA+ScAA6aSZNhTGADylKA5BIwAnKYRg2YFvQdeSnxAK439u-QaSAG4QMADuiCrqChrugZJYyhYAJhBBFhRx8QlJpsS2MBgZ2vQQaDJexJEwmpJ5ipIw+Pk2AOYwxPhBGFBeMOx1MGaqmQHZOSlpGVnjiRYARpXE9MXA9NRQENQA1pFqtTJJTsk2GGEAhBaqAxYbW9tXM4GqTy8BrHofaHpY+Fi0XgYymStGoXhwDHw81oyQAnhRpHJVJQQHAYLBqMQIPQEDwAJyIAAcbA4IEwODw+GocAENHojGYPDYAF0qJs0NtcagyVw8FgIMRCDZoCifOQeCRiDI4IgAPSywEybatKm0LCy-mC4VQAACAEZ8AaAGwagVC6D4fgo4iwmTcVHUYUyUTsTgUnjzGmimzikCS6VyhXlZWq9WeuD6-AAVkNsvDltpNrteDgjogzpZrCAA), you can redraw without worrying about losing input values. Also, redraws are pretty cheap, so unless you're going to have lots of notifications, you don't need to worry about the performance either.

**Comment by cavemansspa** [comment]
@Niklas81 re:
>  thereby not resetting e.g. textareas that users have started filling.

i'm thinking maybe we should to take a step back and understand how you're app / components / state are set up.  as @osban noted, a typical mithril app is resilient to redraws (i.e. your inputs should not be getting reset.)

**Comment by Niklas81** [comment]
Oddly, `oninput: e => text = e.target.value` locks down the text so I can't edit it att all.. I'm not very comfortable with arrow functions, what's that in old school code?
I noticed the Mithril redraws on clicks, which resets my values.. Is that what `oninput: e => text = e.target.value` is supposed to fix? When the event is sent in the background (or through console) it doesn't reset the input boxes anyway. But user must be able to click on other buttons without reseting the input fields.. what's going on?

**Comment by osban** [comment]
@Niklas81 In my code?
Pre-ES6 -> `oninput: function(e) {text = e.target.value}`.
Maybe you can link a flems with your code?

**Comment by isiahmeadows** [comment]
`oninput: e => text = e.target.value` is equivalent to `oninput: function
(e) { return text = e.target.value }`, if that helps.

-----

Isiah Meadows
contact@isiahmeadows.com
www.isiahmeadows.com

**Comment by Niklas81** [comment]
Just for testing this, I had `let test = 'some text` within the view function before return. Putting it outside of it (globally) made it work. :smile:
I might not need independent components after all. Your `oninput` seems to take care of keeping input fields safe from redraws.

However, then I would need to know, how to attach an `addEventListener` to a regular `m('div#someId')`?

**Comment by Niklas81** [comment]
Sorry, got it.. time for bed I think!
Thanks for all the help!

**Comment by spacejack** [comment]
@Niklas81 if you want some (rough) metrics, here are some numbers I consider before I consider trying to optimize full vdom redraws:

* Is the redraw happening very rapidly, like in a requestAnimationFrame loop or drag&drop animation - i.e. many redraws per second.
* Is the vdom very complex (eg. 500-2000 total nodes or more, depending on the target device.) In my experience full redraw slowdowns are perceptible on very low end Android phones once the vdom gets up to around 1000 nodes (multiplied by any heavy computing in your app to build views.)

**Comment by osban** [comment]
@Niklas81 You're welcome :) If you have more questions, you can always ask them in the [Mithril Gitter channel](https://gitter.im/mithriljs/mithril.js).
